scipy/signal/fir_filter_design.py

```diff
@@ -594,3 +603,182 @@ def remez(numtaps, bands, desired, weight=None, Hz=1, type='bandpass',
     bands = np.asarray(bands).copy()
     return sigtools._remez(numtaps, bands, desired, weight, tnum, Hz,
                            maxiter, grid_density)
+
+
+def firls(numtaps, bands, desired, weight=None, nyq=1.):
+    """
+    FIR filter design using least-squares error minimization.
+
+    Calculate the filter coefficients for the linear-phase finite
+    impulse response (FIR) filter which has the best approximation
+    to the desired frequency response described by `bands` and
+    `desired` in the least squares sense (i.e., the integral of the
+    weighted mean-squared error within the specified bands is
+    minimized).
+
+    Parameters
+    ----------
+    numtaps : int
+        The number of taps in the FIR filter.  `numtaps` must be odd.
+    bands : array_like
+        A monotonic nondecreasing sequence containing the band edges in
+        Hz. All elements must be non-negative and less than or equal to
+        the Nyquist frequency given by `nyq`.
+    desired : array_like
+        A sequence the same size as `bands` containing the desired gain
+        at the start and end point of each band.
+    weight : array_like, optional
+        A relative weighting to give to each band region when solving
+        the least squares problem. `weight` has to be half the size of
+        `bands`.
+    nyq : float, optional
+        Nyquist frequency. Each frequency in freq must be between 0 and
+        `nyq` (inclusive).
+
+    Returns
+    -------
+    coeffs : ndarray
+        Coefficients of the optimal (in a least squares sense) FIR filter.
+
+    See also
+    --------
+    scipy.signal.firwin
+    scipy.signal.firwin2
+
+    Examples
+    --------
+    We want to construct a band-pass filter. Note that the behavior in the
+    frequency ranges between our stop bands and pass bands is unspecified,
+    and thus may overshoot depending on the parameters of our filter:
+
+    >>> from scipy import signal
+    >>> import matplotlib.pyplot as plt
+    >>> fig, axs = plt.subplots(2)
+    >>> nyq = 5.  # Hz
+    >>> desired = (0, 0, 1, 1, 0, 0)
+    >>> for bi, bands in enumerate(((0, 1, 2, 3, 4, 5), (0, 1, 2, 4, 4.45, 5))):
+    ...     fir_firls = signal.firls(73, bands, desired, nyq=nyq)
+    ...     fir_remez = signal.remez(73, bands, desired[::2], Hz=2 * nyq)
+    ...     fir_firwin2 = signal.firwin2(73, bands, desired, nyq=nyq)
+    ...     hs = list()
+    ...     ax = axs[bi]
+    ...     for fir in (fir_firls, fir_remez, fir_firwin2):
+    ...         freq, response = signal.freqz(fir)
+    ...         hs.append(ax.semilogy(nyq*freq/(np.pi), np.abs(response))[0])
+    ...     for band, gains in zip(zip(bands[::2], bands[1::2]), zip(desired[::2], desired[1::2])):
+    ...         ax.semilogy(band, np.maximum(gains, 1e-7), 'k--', linewidth=2)
+    ...     ax.legend(hs, ('firls', 'remez', 'firwin2'), loc='lower center')
+    ...
+    >>> plt.show()
+
+    Notes
```

Comment:
the order of sections in this docstring is off, should be Notes-References-Examples (see https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt)